qubes-rpc/qubes.StartApp

```diff
@@ -1,25 +1,50 @@
-#!/bin/sh
+#!/usr/bin/python3 --
 
-if [ -z "$1" ]; then
-    echo "This service requires an argument" >&2
-    exit 1
-fi
+import sys, os, pwd
+import qubesdb
+from qubesagent.xdg import launch
 
-app_basename="$1.desktop"
+def main(argv):
+    try:
+        (arg0, arg,) = argv
+    except ValueError:
+        print('This service requires an argument', file=sys.stderr)
+        sys.exit(1)
+    app_name = os.path.join('applications', arg + '.desktop')
+    try:
+        xdg_data_home = os.environ['XDG_DATA_HOME']
+    except KeyError:
+        xdg_data_home = os.path.join(pwd.getpwuid(os.getuid()).pw_dir, '.local/share')
+    try:
+        xdg_data_dirs = os.environ['XDG_DATA_DIRS']
+    except KeyError:
+        xdg_data_dirs = ['/usr/local/share','/usr/share']
+    else:
+        if '::' in xdg_data_dirs or \
+                xdg_data_dirs.startswith(':') or \
+                xdg_data_dirs.endswith(':'):
+            print('Empty entries in XDG_DATA_DIRS not allowed')
+            sys.exit(1)
+        xdg_data_dirs = xdg_data_dirs.split(':')
+    paths_to_try = [xdg_data_home] + xdg_data_dirs
+    qubes_db = qubesdb.QubesDB()
+    if qubes_db.read('/qubes-service/app-dispvm.' + arg) is not None:
+        try:
+            del os.environ['XDG_DATA_HOME']
+        except KeyError:
+            pass
+        os.environ['XDG_DATA_DIRS'] = '/usr/share/qubes/mime'
+    for path in paths_to_try:
+        desktop_file = os.path.join(path, app_name)
```

Comment:
This alone won't be enough, you need to check also subdirectories - see `find ... -type d` call in the original code. This applies to some KDE applications.
On a second thought, we might have encoded application ID wrong - see https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id, specifically `turn '/' into '-'` part. Currently we take only the last path part. 
And also, the specification is explicitly ambiguous: `If both foo-bar.desktop and foo/bar.desktop exist, it is undefined which is selected`.